### ISSUE TITLE
Consider swapping busboy for multer

### ISSUE THREAD

**Issue by ErisDS** [TryGhost/Ghost#6362]
It would be good to see an investigation into the viability of swapping [busboy](https://github.com/mscdex/busboy) for [multer](https://github.com/expressjs/multer). Busboy has a complex API and requires us to have quite a lot of our own code, multer is simpler to use.

Multer is also under the expressjs repo, which lends it some extra credence. However, I have no idea which is least buggy or most actively maintained, or if multer is a suitable replacement for our use cases.

It would be good to look into both properly, investigate and try out the switch, and see if it's a worthwhile change.

**Comment by edsadr** [comment]
Multer is just another layer on top of Busboy, the abstraction is way simpler, but the project will still depend on Busboy at the end.

Multer is pretty good shape and actively maintained: https://github.com/expressjs/multer/graphs/contributors

The switch in Ghost doesn't look hard to accomplish as far as I can see, is a matter of adding the dependency and implementing another middleware to replace `core/server/middleware/ghost-busboy.js` which basically is used for post requests...

I am willing to try helping with this one if you are ok @ErisDS

**Comment by ErisDS** [comment]
I think if we can simplify the code in Ghost core that's a boon :+1: 

Would also be interesting to know if swapping would resolve https://github.com/TryGhost/Ghost/issues/6340

**Comment by edsadr** [comment]
Ok I will start with this... will let you know

**Comment by ckousik** [comment]
pr #6404
Wrote a new middleware like @edsadr suggested. Works fine on localhost.

**Comment by ErisDS** [comment]
Seeing the implementation in PR #6404 makes me feel this may not be worth while. It doesn't get rid of much code, but leaves me with a lot of questions:

- is the complex filename function necessary?
- isn't the os tmpdir the default place to upload files for multer?
- how are the errors handled?
- is there any difference in performance or reliability?

The PR says that its trying to mimic the existing behaviour, but is that the right thing to do?

Essentially, this issue was raised looking for an investigation and/or some knowledge/experience from the community on whether the swap would provide benefits. The PR provided includes an example implementation but I'm looking for some information / testing to help inform a decision.

**Comment by franciscop** [comment]
What about [formidable](https://github.com/felixge/node-formidable)? I just tried it yesterday and it also seems great for this.

**Comment by ErisDS** [comment]
> What about formidable? I just tried it yesterday and it also seems great for this.

It would be great if you could provide some reasoning. Why is it great? Why would it be a good replacement?

**Comment by franciscop** [comment]
TL;DR: The main points for it compared to multer are that the API is even simpler and it's been battle tested. The main point against it is that it doesn't seem to be in active development any longer.

Compared to busboy, you don't have to handle each step of the process as in `file.on('data' ...)` or `file.on('end' ...)` so it's just a function with a callback:

```js
exports.upload = function(req, res){
  var form = new formidable.IncomingForm();
  form.parse(req, function(err, fields, files) {
    // ...
  });
};
```

But as I'm seeing the *[multer pull request](https://github.com/TryGhost/Ghost/pull/6404/files)*, that seems to be in a similar level of abstraction and simplicity. *Formidable* also seems to be quite battle-tested:

> This module was developed for Transloadit, a service focused on uploading and encoding images and videos. It has been battle-tested against hundreds of GB of file uploads from a large variety of clients and is considered production-ready.

However, the issues are piling up [and not being frequently closed](https://github.com/felixge/node-formidable/issues?q=is%3Aissue+is%3Aclosed), and the last significant code [was written in 2013](https://github.com/felixge/node-formidable/graphs/contributors). Multer has 2 years lifespan but support from expressjs.

**Comment by ErisDS** [comment]
Would love to hear some discussion around my questions here: https://github.com/TryGhost/Ghost/issues/6362#issuecomment-179309604

It seems to me that if we are going to switch, multer is a better choice than formidable due to activity. However, I'm still interested in some discussion around the implementation & particularly whether our current behaviour should be mimicked or improved.

**Comment by acburdine** [comment]
@ErisDS can this be closed?

**Comment by ErisDS** [comment]
Aye this was closed by #6659 / c41ee35
